### ISSUE TITLE
IMPORTANT: DO NOT USE! Use import mapping instead

### ISSUE THREAD

**Issue by Nytelife26** [ilearnio/module-alias#113]
A similar feature to this library is already available in node as a standard: [import mapping](https://nodejs.org/api/packages.html#packages_imports).

Not only does it allow for directory mapping, but it also allows dependency aliasing, and works for both `require` and `import` (ESM) WITHOUT breaking resolution behaviour in production or other libraries like using this library does.

You have been warned.

I know this sounds hateful, however, that is not my intention. My intention is instead to spread awareness about this mostly undocumented feature (it is not shown in any package manifest documentation besides node's, and it is relatively tucked away) so people can write better software without needing to use hacky libraries like this one and without adding unnecessary dependencies.

**Comment by Kehrlann** [comment]
@Nytelife26 , thanks for raising awareness. This was introduced "recently", so on older versions of node, you might still want `module-alias` - therefore we probably won't just "deprecate" the package.

Would you mind opening a PR to update the README of this package, explaining that import mapping exists and should be use instead?

**Comment by Nytelife26** [comment]
Yes, that is fine. I would be more than happy to open a pull request regarding
versioning and use of import mapping where possible. Thank you for the response.

**Comment by nick-bull** [comment]
Further reading [here](https://nodejs.org/api/packages.html#packages_subpath_imports) and with a great example of directory mapping [here](https://nodejs.org/api/packages.html#packages_subpath_patterns). Hopefully this latter example clears up any confusion about the difference between `"imports"` and `"exports"` usage too

**Comment by Nytelife26** [comment]
@nick-bull Thanks for the extra examples :) greatly appreciated

**Comment by nick-bull** [comment]
An interest afterthought to replacing this module is that import mapping does not resolve directory imports, e.g. `"#services/*": "./src/services/*.js"` would not resolve `import ... from '#services/someCoolService'` to  `./src/services/someCoolService/index.js`.

You can get around this with `"#services/*": "./src/services/*/index.js"` but then that obviously doesn't work with non-`index.js` mappings, and the value is not really a regex, so something like `"./src/services/*(/index)?.js"` would not work. Directory-style imports with `"#services/*": "./src/services/*.js"`, `import ... from '#services/someCoolService` throw `Error [ERR_UNSUPPORTED_DIR_IMPORT]: Directory import ... is not supported resolving ES modules imported from ...`

Anybody know a way around this?

**Comment by Nytelife26** [comment]
@nick-bull ..uh
```
"#services/*": "./src/services/*.js"
```
Of course it doesn't? That'd resolve to `./src/services/someCoolService.js"
Use this instead:
```
"#services/*": "./src/services/*"
```

**Comment by nick-bull** [comment]
> @nick-bull ..uh
> 
> ```
> "#services/*": "./src/services/*.js"
> ```
> 
> Of course it doesn't? That'd resolve to `./src/services/someCoolService.js`
> Use this instead:
> 
> ```
> "#services/*": "./src/services/*"
> ```

I have already suggested that in my comment; I've edited my response so it's a little clearer, as it was a bit densely packed, as it does not work, it throws an error

**Comment by Nytelife26** [comment]
@nick-bull No, you did not already suggest what I said specifically. You suggested something similar, which was `"#services/*": "./src/services/*.js"` as aforementioned. Although, actually, that still works - you'll notice if you import `#services/someCoolService/test`, that does in fact resolve to `./services/someCoolService/test.js`, for both require and import. And then if you do the other thing I suggested, `"#services/*": "./src/services/*"`, you have to import `#services/someCoolService/test.js` to achieve the same result, but it will definitely work.

I hope that helps.

**Comment by nick-bull** [comment]
> @nick-bull No, you did not already suggest what I said specifically. You suggested something similar, which was `"#services/*": "./src/services/*.js"` as aforementioned. Although, actually, that still works - you'll notice if you import `#services/someCoolService/test`, that does in fact resolve to `./services/someCoolService/test.js`, for both require and import. And then if you do the other thing I suggested, `"#services/*": "./src/services/*"`, you have to import `#services/someCoolService/test.js` to achieve the same result, but it will definitely work.
> 
> I hope that helps.

You're right, I totally thought I'd written that example too and forgot to include it. I'm still not sure that answers my question though - is there a way to write a mapping that will allow both of the following:

```
import ... from '#coolService' // coolService/index.js
import ... from '#coolService/someFile.js' // coolService/someFile.js
```

**Comment by Nytelife26** [comment]
No worries @nick-bull. My answer if not for your discovery of needing to set the type to modules would've been to just import `#coolService/index.js` if it wouldn't work the conventional way anyways, but I'm glad you found your problem.

If you have any further questions or concerns let me know.

**Comment by thefoxie** [comment]
This package didn't  work for me no matter how I hard I tried to configure it. Maybe because I use ES6+ including imports (no single require() in my project) Builtin functionality works.
```json
"imports": {
    "##/*": "./*"
},
```
jsconfig
```json
{
    "compilerOptions": {
        "baseUrl": ".",
        "paths": {
            "##/*": ["./*"]
        }
    },
    "exclude": ["node_modules"]
}

```
usage
```js
import { Model } from "##/api/models.js";
```

**Comment by Nytelife26** [comment]
@thefoxie thank you for the contribution :) however, this is not a place to report the package failing. I am glad the built-in functionality works, however, if you have any problems you should still create a separate issue so the ilearnio team can ensure the package works as intended.

**Comment by Papooch** [comment]
After reading the documentatino, I still don't understand how to use it... Could you please explain to me how can I convert this:
```
"_moduleAliases": {
    "@root": ".",
    "@submodules": "submodules",
    "@db": "src/db",
    "@middleware": "src/middleware"
  }
```
to use native import mapping instead? Preferably without changing any other code. Thanks.

**Comment by Nytelife26** [comment]
@Papooch you *will* have to convert to using `#alias` instead of `@alias`, but that's it.
```json
"imports": {
    "#root/*": "./*",
    "#submodules/*": "./submodules/*",
    "#db/*": "./src/db/*",
    "#middleware/*": "./src/middleware/*"
}
```
I hope that helps.

**Comment by Papooch** [comment]
Well that's what I tried also, but I am getting and error.

in `C:\repos\project\src\db\connect.js`, I am using
```js
const { initModels } = require('#submodules/db-models/master');
```
But it throws an error, which I don't understand:
```
Error: Cannot find module 'C:\repos\project\submodules\db-models\master'
    at Object.<anonymous> (C:\repos\project\src\db\connect.js)
```
But the folder `C:\repos\project\submodules\db-models\master` DOES exist and has an `index.js` file in it which exports the initModels function.

Does folder mapping work differently to file mapping using this technique?

**Comment by nick-bull** [comment]
@Papooch You didn't happen to get burned in [the same way as I did](https://github.com/ilearnio/module-alias/issues/113#issuecomment-784247113)? Test by using 

```
const { initModels } = require('#submodules/db-models/master/index.js');
```

**Comment by Papooch** [comment]
@nick-bull Well, that does seem to work, but I have to explicitly state the file, including the file extension. `require('#submodules/db-models/master/index')` does *not* work. I would have to change imports in the whole project and lose the flexibility along the way.

I would very much prefer to use a native solution instead of a library, but not in a way that makes my experience worse.

**Comment by nick-bull** [comment]
@Papooch I suspect you're being burned for the same reason as the prior mentioned comment, this feature isn't going to affect extension resolution. Try `node --es-module-specifier-resolution=node`, or install `esm` and use `node -r esm`. Report back and let us know if that works!

**Comment by initplatform** [comment]
I've gotten this to work as described above.

In my package.json

```json
    "imports": {
        "#app/*": "./dist/app/*",
        "#lib/*": "./dist/lib/*",
        "#src/*": "./dist/*"
    },
```

then using `node --es-module-specifier-resolution=node` after compilation.

Where I am having issues, is when I try and use jest. Previously when using module-alias I could do the following in jest.config.ts

```typescript
    moduleNameMapper: {
        '^#app/(.*)$': '<rootDir>/src/app/$1',
        '^#lib/(.*)$': '<rootDir>/src/lib/$1',
        '^#src/(.*)$': '<rootDir>/src/$1'
    },
``` 

but when I try and run tests, now, I am getting:


```
    Configuration error:

    Could not locate module #app/index.js mapped as:
    /Users/blah/src/app/$1.

    Please check your configuration for these entries:
    {
      "moduleNameMapper": {
        "/^#app\/(.*)$/": "/Users/blah/src/app/$1"
      },
      "resolver": undefined
    }

    > 1 | import { FastifyApp } from '#app/index.js';
```

It looks like it is looking for `#app/index.js` instead of `#app/index.ts`

If I do a quick hardcode hack, the test runs (I hardcode index.ts):

```typescript
    moduleNameMapper: {
        '^#app/(.*)$': '<rootDir>/src/app/index.ts',
        '^#lib/(.*)$': '<rootDir>/src/lib/$1',
        '^#src/(.*)$': '<rootDir>/src/$1'
    },
```

...any ideas?

**Comment by Nytelife26** [comment]
@Papooch if you wish to use extensions in the map to avoid having to specify `.js` you can. Although, as people have mentioned, it is better to set the module resolution to use node's. 

@initplatform how are the imports specified in the actual files?

**Comment by initplatform** [comment]
Ha!
I was banging my head against the monitor for a good hour earlier before I wrote that post. 
5 minutes after trying to respond to your question I found it. Thanks! 

For whatever reason, when I was debugging the transition over to import mapping I had done this:

```typescript
import { FastifyApp } from '#app/index.js';
```

instead of this:

```typescript
import { FastifyApp } from '#app/index';
```

Can't even remember what I was testing... but I forgot to remove the `.js` after whatever I was doing.
Building and running worked fine, but jest exposed the error. 

Sometimes it's the tiny things right :)

I am excited to see this import mapping land in node. At first I was averse to the `#` instead of using the `@`, but I kinda like it now. I am assuming it's to differentiate between npm namespaces and actual mappings...

Thanks to all involved that created `module-alias` and for your response @Nytelife26

**Comment by nick-bull** [comment]
@initplatform I'm not sure you'll need the `index` either, as I'm pretty sure directory imports are part of `--es-module-specifier-resolution=node`

**Comment by Nytelife26** [comment]
No worries @initplatform. I give special credit to @nick-bull for the module resolution override because prior to that I was unaware it even existed honestly. I'm glad you found your solution, though!

**Comment by initplatform** [comment]
> @initplatform I'm not sure you'll need the `index` either, as I'm pretty sure directory imports are part of `--es-module-specifier-resolution=node`

I believe that's an annoyance with typescript.

For whatever reason, if only using the typescript path alias, I need to specify index or typescript won't compile. I don't need to use index for nested routes though. 

<img width="471" alt="Screen Shot 2021-03-15 at 9 52 45 AM" src="https://user-images.githubusercontent.com/47506789/111164482-80e31b80-8574-11eb-9a5d-192f01a868a8.png">

<img width="480" alt="Screen Shot 2021-03-15 at 9 55 44 AM" src="https://user-images.githubusercontent.com/47506789/111164675-ab34d900-8574-11eb-9347-61ebc2d6b522.png">

**Comment by nick-bull** [comment]
One other reason to keep maintaining this package is that import mapping only works for `"type": "module"`, which isn't ideal for e.g., React Native

**Comment by Nytelife26** [comment]
> One other reason to keep maintaining this package is that import mapping only works for `"type": "module"`, which isn't ideal for e.g., React Native

Is that so? Standard React and JSX relies on ESM, I cannot see why React Native wouldn't permit it.

**Comment by Griffork** [comment]
I've discovered that I might need to switch to es6 modules, and that module-alias seems to be not working. I stumbled across this thread and converted my project over to using the "imports" property in package.json, but I'm getting an error I can't solve:

Package.json:
```json
"imports": {
    "#Common/*": "../Common/*.js",
    "#Typescript/*":  "../Typescript/*.js"
  }
```

Error:
```txt
Invalid "imports" target "../Common/*.js" defined for '#Common/*' in the package config [PathToRepo]\Server\package.json imported from [PathToRepo]\Server\server.js
```

Does this mean that the inbuilt "imports" property doesn't work outside the directory that the package.json is located within? Is there a work around for this? If not is there a way to make module-alias work with ES6 modules?

**Comment by Nytelife26** [comment]
> Does this mean that the inbuilt "imports" property doesn't work outside
> the directory that the package.json is located within?

Do I get to ask why you're trying to use import paths to import from
directories not contained within your project? The paths you're aliasing
should *always* be contained within the project that the `package.json`
is for.

It appears that you're working on a project with multiple subpackages.
If that is the case, use the import paths in the `package.json` for the
top level directory of the project.

I hope this helps.

**Comment by rayfoss** [comment]
I can't figureout how to get node 16 native ESM support to work with Mocha 8 and module-alias... So this is my only option

**Comment by Griffork** [comment]
@Nytelife26 because I have the following directory structure:

  - Client
  - Common
  - Server
   
Client and server both include common, so it can't be "within" either project.

Also WHY do you need to enforce a very arbitrary condition of "you can not alias any code outside of the directory of your project'? Shared node_modules folders, shared code and external libraries being outside of your codebase (for the purposes of cleaner commits to git repositories and such) are not uncommon occurances in the wild. 

My top level doesn't have a package.json (because it's not a package ...?). Having package.jsons that exist alone with no actual code associated with them seems like a bad design choice. 

Instead I hacked an experimental node API to allow accessing paths outside of the current directory.

**Comment by Nytelife26** [comment]
@Griffork That seems to be a relatively critical misunderstanding of how monorepositories work. They're packages within a package structure.

Common is not part of either package. So, instead, you should make the other packages *depend* on common, assuming common is its own package.

As to why that needs to be a "very arbitrary condition", how do you expect a package to work if it imports code that is neither bundled with the package nor depended on in the manifest?? If anyone installs that, it ***will*** break, guaranteed. Shared code and external libraries are common - that's why you can make them dependencies.

You should look into workspaces for your relevant node package manager and research further into monorepositories.

**Comment by prabhatCH** [comment]
> A similar feature to this library is already available in node as a standard: [import mapping](https://nodejs.org/api/packages.html#packages_imports).
> 
> Not only does it allow for directory mapping, but it also allows dependency aliasing, and works for both `require` and `import` (ESM) WITHOUT breaking resolution behaviour in production or other libraries like using this library does.
> 
> You have been warned.
> 
> I know this sounds hateful, however, that is not my intention. My intention is instead to spread awareness about this mostly undocumented feature (it is not shown in any package manifest documentation besides node's, and it is relatively tucked away) so people can write better software without needing to use hacky libraries like this one and without adding unnecessary dependencies.

However, this works fine for me. VS Code Intellisense will not resolve the imports. Is there any convenient workaround for that?

**Comment by Nytelife26** [comment]
> this works fine for me

Something working does not mean it works well, nor that you should use
it.

> VS Code Intellisense will not resolve the imports. Is there any convenient
> workaround for that?

That seems to be an issue with current tooling and import resolution.
ESLint with the `node/no-unresolved-imports` rule fails in the same way.
Hopefully soon more people will implement checking for this - including
language servers - and then it will work better.

**Comment by cnm-n** [comment]
@Nytelife26 Trying to switch to a node subpath imports, but for me it only works with the full path.

My imports:
```
  "imports": {
    "#utils/*": "./app/utils/*"
  }
```
Utils folder has two files: `index.js` and `test.js`, both has `foo` and `bar` functions.

My entry: 
```
const {foo} = require('#utils');
const {bar} = require('#utils/test');
foo();
bar();
```
Both calls only work if I add the full path: `#utils/index.js` or `#utils/test.js`. Otherwise i have:
```
Package import specifier "#utils" is not defined in package C:\...\package.json imported from C:\...\test.js
and
Cannot find module 'C:\Users\...\app\utils\test'
```
I have no `"type": "module"` enabled, and I tried `"type": "commonjs"` with no success. Also `experimental-specifier-resolution` does not help.

Any idea what I'm doing wrong?

**Comment by Spence-S** [comment]
To see a working example of using subpath imports with `index.js` files - see the [forwardemail.net repo ](https://github.com/forwardemail/forwardemail.net/blob/master/package.json#L234)

package.json
```json
  "imports": {
    "#config/*": "./config/*.js",
    "#config": "./config/index.js",
    "#helpers/*": "./helpers/*.js",
    "#models/*": "./app/models/*.js",
    "#models": "./app/models/index.js",
    "#controllers/*": "./app/controllers/*.js",
    "#controllers": "./app/controllers/index.js",
    "#controllers/web": "./app/controllers/web/index.js",
    "#controllers/api": "./app/controllers/api/index.js"
  },
```

Many people are confused because using these import maps uses the same resolution logic as ESModules, which do not automatically resolve index.js files in the same way commonjs does. ESM requires the full path and extension to the file it is importing, and by extension so do the import maps here. 

Using import maps like shown above will help you have a similar feel to commonjs. If you go deeper in the linked repo - you can see that this works. 

Also note that the `*` in the imports does not have the same functionality as a glob star, which can be a bit confusing and may need some trial and error to set up your project correctly.

**Comment by viT-1** [comment]
Best solution to solve aliases/paths in js projects for node scripts is using [tsconfig-paths](https://github.com/dividab/tsconfig-paths). For [example](https://github.com/viT-1/dist-gh-pages/commit/8fca781dede871be8f8f5d7841f00f3424878b34).
Jest & Eslint have abilities to use tsconfig.json paths too.

**Comment by viT-1** [comment]
> Many people are confused because using these import maps uses the same resolution logic as ESModules, which do not automatically resolve index.js files in the same way commonjs does. ESM requires the full path and extension to the file it is importing, and by extension so do the import maps here.

@Spence-S Yeah, I use vanilla nodejs, without bundlers, and I'm forced to use replace-in-file [solution](https://github.com/viT-1/dist-gh-pages/blob/8fca781dede871be8f8f5d7841f00f3424878b34/replace-in-file.js#L19) with regexp for solving extension & index addition.
 
> Using import maps like shown above will help you have a similar feel to commonjs. If you go deeper in the linked repo - you can see that this works.
How about glob paths? Something like that `src/**/*`

**Comment by viT-1** [comment]
Another solution is [tsc-alias](https://github.com/justkey007/tsc-alias) which generates relative paths.
You can see [my configuration](https://github.com/viT-1/dist-gh-pages/commit/8ccbffd4531564dd5332e2d8e1ec3238a5e81a49) for it.

**Comment by eng-dibo** [comment]
> A similar feature to this library is already available in node as a standard: [import mapping](https://nodejs.org/api/packages.html#packages_imports).
> 
> Not only does it allow for directory mapping, but it also allows dependency aliasing, and works for both `require` and `import` (ESM) WITHOUT breaking resolution behaviour in production or other libraries like using this library does.
> 
> You have been warned.
> 
> I know this sounds hateful, however, that is not my intention. My intention is instead to spread awareness about this mostly undocumented feature (it is not shown in any package manifest documentation besides node's, and it is relatively tucked away) so people can write better software without needing to use hacky libraries like this one and without adding unnecessary dependencies.

unfortunately, the native imports mapping forces you to rewrite all of your path aliases to be start with '#' which breaks your codebase

**Comment by eng-dibo** [comment]
> Best solution to solve aliases/paths in js projects for node scripts is using [tsconfig-paths](https://github.com/dividab/tsconfig-paths). For [example](https://github.com/viT-1/dist-gh-pages/commit/8fca781dede871be8f8f5d7841f00f3424878b34). Jest & Eslint have abilities to use tsconfig.json paths too.

this doesn't work if you set `type: module` in your `package.json`

**Comment by eng-dibo** [comment]
this will force you make some changes to your codebase that introduce breaking change

you will need to change your alias to start with '#'
es6 imports can't refer to a path outside your module i.e a parent director for monorepos

**Comment by jamesmortensen** [comment]
For those who say changing '@' to '#' will break their codebase, is it possible for you to do a find/replace in your codebase to fix the imports?  We've done something like that before when needing to do other sweeping changes across the codebase, and it's worked out quite well as long as there was a pattern we could match on initially.

**Comment by eng-dibo** [comment]
> For those who say changing '@' to '#' will break their codebase, is it possible for you to do a find/replace in your codebase to fix the imports? We've done something like that before when needing to do other sweeping changes across the codebase, and it's worked out quite well as long as there was a pattern we could match on initially.

 the problem is not how to replace '@' by '#', the problem is in the project structure itself.

for example if you want to import a local package that is not published to npm and updated frequently without packing it everytime
you can add it's name to tsconfig path so `@scope/package-name` refers the location of the local package
and allows you to import it in your project like this
```
import something from `@scope/package-name
```
which will be replaced with the actual path.

changing '@' to '#' will not help

**Comment by eurostar-fennec-cooper** [comment]
One interesting thing I've come across recently when trying to run both tsc->node for the build/start script and ts-node for the dev script is that neither module-alias nor node.js' new import mapping support aliasing paths _relative to the working directory_.

For example, my src folder contains Typescript files. Package.json contains a _moduleAliases alias from "@" to "dist", which is the folder typescript will build to. This works perfectly fine, the start script is essentially `node dist/index`.

The issue occurs when running something like ts-node, the dev script is: `ts-node src/index`. However because module-alias is configured to alias to dist, at runtime ts-node cannot find modules as it's looking in the dist folder, but is operating out of the src folder.

In the end it was solved by using both module-alias and tsconfig-paths, the former using alias/paths config from package.json when running from dist/build folder. And the latter using alias/paths config from tsconfig.json when running from the src folder.

All of this module/path aliasing stuff makes things so much tidier, but all implementations seem to rely on aliases to a static path. Imagine if for some reason I needed two separate copies of my code in different src or dist/build folders, which current path aliasing solutions it's impossible; imports in "src2" would reference files in the src folder.

**Comment by jlenon7** [comment]
@spence-s when you use the Node.js ESM import aliases in your code to import modules, does VSCode understand the type of the class or function that you are importing? In my case I can only get the type of the class or function if I use the real path to import modules:

```js
// Does not work. VSCode does not show intellisense for Hello class.
import { Hello } from '#src/hello'

// Works
import { Hello } from './src/Hello.js'
```

**Comment by Spacey4K** [comment]
Would it be possible to resolve the path from subpath Imports?
Similar functionality as #33  

For example if i have this import for assets:
```json
"imports": {
    "#config": "./src/config.json",
}
```

How would i resolve the path to a variable?
Something as:
```js
const path = getAliasPath("#config");
console.log(path); // "./src/config.json" or absolute path
```

**Comment by spence-s** [comment]
@jlenon7 Yeah I don't think the path intellisense extensions support subpath imports, however, TS (and vscode ts linting) does. So while it won't help you autocomplete, it will help validate their existence and types.

**Comment by jamesmortensen** [comment]
@TheDirigible why not make Common into its own package so it can be added as a dependency?  You of course can setup a project however you like. No one is trying to force anything, but I think the idea here is that NPM has features and functionality to include files and libs from outside the project, so that installing it is a one command operation.  If it's possible to make those changes in the project to set it up to work with the existing tooling, then configuration and managing it could potentially be a lot easier.

**Comment by Nytelife26** [comment]
@eng-dibo 

Local dependencies that aren't published to NPM are registered like this.
```json
dependencies: {
    "@some/package": "file:./path/to/package",
}
```
I'm not sure what makes you think you have to update your code for this, nor why you're using import aliases for local packages when the file dependency specifier has existed for quite some time now.

**Comment by Nytelife26** [comment]
> And what if Common is not it's own package?

Common was very clearly a package of some description in their example, seemingly as a shared library.

Aside from making everything you want to use an actual dependency, there is no solution that makes sense.
In fact, I'm not even sure why your structure exists. If you were to make the library a local package, you can import things from it in as many places as you want without needing to symlink or create aliases. Life makes more sense that way.

**Comment by deakjahn** [comment]
@Nytelife26 I'm struggling with the same, and this is the first time I do this, so I might be mistaken, but I spent the better part of two days with this. It looks like the local dependency you mention has the significant drawback that the building process will _not_ build the dependent project even if needed, It looks like the only way to make TSC build it automatically is to use [project references](https://www.typescriptlang.org/docs/handbook/project-references.html).

**Comment by deakjahn** [comment]
I kept going on and I think I have the solution, without symlinks. It's rather fragile, in the sense that is needs a very carefully balanced folder structure to work.

I'll keep the client-server-common paradigm for the discussion, both the _client_ and the _server_ depending on the _common_ project. Let's start with the topmost, "solution" folder.

Apart from other settings, it needs a `tsconfig.json` with -- to make sure TSC builds both dependent projects:

    "references": [
      {
        "path": "./client"
      },
      {
        "path": "./server"
      },
     ]

And a `package.json` with -- to make sure the running code will be able to load the common module:

    "imports": {
      "#common/foobar": "./out/common/src/foobar.d.ts"
    },

Both the _client_ and the _server_ packages need a `tsconfig` with -- to make sure that they get output to a common `out` folder, to help TS (IDE) to find the common module and to depend on it for the build process:

    "compilerOptions": {
      "baseUrl": ".",
      "rootDir": "../",
      "outDir": "../out",
      "sourceMap": true,
      "strict": true,
      "paths": {
        "#common/*": [
          "../common/src/*.ts"
        ]
      },
    },
    "references": [
      {
        "path": "../common"
      }
    ],

The _common_ project needs a `tsconfig` with -- also to output to the same folder, and to act as an appropriate `TSC --build` dependent project:

    "compilerOptions": {
      "baseUrl": ".",
      "rootDir": "../",
      "outDir": "../out",
      "composite": true,
      "declaration": true,
      "declarationMap": true,
    },
    "references": []

I found that if the `outDir` or `rootDir` folders have any other structure, the whole scheme falls apart.

External dependencies are left for the specific case, each subproject needs its own dependencies and the topmost `package.json` will also need the relevant packages in its `node_modules`.

**Comment by strblr** [comment]
Life on easy mode: use [tsx](https://github.com/esbuild-kit/tsx) as a runtime instead of node. Everything just works, no need for additional config besides the aliases defined in `tsconfig.json`.

**Comment by KocakAli** [comment]
I don't understand what I am missing. I am getting this error:

Package import specifier "#utils/math" is not defined in package /Users/alikocak/typetest/package.json imported from /Users/alikocak/typetest/dist/main.js

I am using import
import { add } from "#utils/math";

tsconfig.json
`{
    "compilerOptions": {
      "module": "ESNext",
      "target": "ESNext",
      "moduleResolution": "node",
      "outDir": "./dist",
      "rootDir": "./src",
      "baseUrl": ".",
      "paths": {
        "#utils/*": ["src/utils/*"]
      }
    }
  }`
  
  package.json
  `{
  "name": "typetest",
  "version": "1.0.0",
  "export": "dist/main.js",
  "type": "module",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "start": "node dist/main.js"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "description": "",
  "imports": {
    "#utils/*": "./dist/utils/*"
  }
}
`

**Comment by jjangga0214** [comment]
@strblr But after compiling ts to js, you still need to resolve how alias should work somehow.

(For folks, check [this](https://github.com/jjangga0214/ts-yarn-lerna-boilerplate?tab=readme-ov-file#module-aliases) out for various options: )

**Comment by KocakAli** [comment]
@jjangga0214 I think first solution subpath importing is enough in most cases. Is there any technical reason not to use subpath importing?

**Comment by Nefcanto** [comment]
I did not have time to read all of this thread. But I wish you had added the top message to your README file, to prevent days of frustration for new developers not seeing results in newer Node version.
